File: crm/agent/first_touch_reply.py
"""Booking SMS after a lead replies to first-touch — autonomous-run backstop.

The TypeScript event trigger (`lib/agent/first-touch-reply.ts`) sends as
soon as the inbound SMS lands. This module does the same job when the
autonomous run drains the trigger queue.

Sends through Telnyx. Status is sent. Confirm the time they picked, or
offer two concrete showing windows. Voice comes from AIUserProfile.
Missing credentials fail — they do not fall back to a pending draft.
"""

from __future__ import annotations

import re
import uuid
from datetime import datetime, timedelta, timezone
from typing import Any

from first_touch import (
    compose_first_touch_sms,
    first_name_of,
    normalize_tone,
    propose_two_showing_windows,
    send_sms,
)

INBOUND_MESSAGE_EVENT = "inbound_message"
_DEDUPE_WINDOW_HOURS = 48
_FIRST_TOUCH_LOOKBACK_DAYS = 14
_FIRST_TOUCH_REASON_MARK = "First-touch SMS"
_REPLY_REASON = "Reply to first-touch — book a showing. Sent."

_WINDOW_LABEL_RE = re.compile(
    r"\b(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)\s+\d{1,2}(?::\d{2})?(?:am|pm)\b",
    re.IGNORECASE,
)
_DAY_ALIASES = {
    "mon": ("mon", "monday"),
    "tue": ("tue", "tues", "tuesday"),
    "wed": ("wed", "wednesday"),
    "thu": ("thu", "thur", "thurs", "thursday"),
    "fri": ("fri", "friday"),
    "sat": ("sat", "saturday"),
    "sun": ("sun", "sunday"),
}
_CONFIRM_MARKS = {
    "warm": ("does that still work",),
    "direct": ("still work for you",),
    "formal": ("still work for you",),
    "casual": ("still good",),
}


def is_inbound_message_event(event: str | None) -> bool:
    return event == INBOUND_MESSAGE_EVENT


def is_first_touch_draft(row: dict[str, Any] | None) -> bool:
    return bool(row) and _FIRST_TOUCH_REASON_MARK in (row.get("reasoning") or "")


def is_first_touch_reply_draft(row: dict[str, Any] | None) -> bool:
    return bool(row) and "Reply to first-touch" in (row.get("reasoning") or "")


def extract_window_labels(text: str) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for match in _WINDOW_LABEL_RE.findall(text or ""):
        if match in seen:
            continue
        seen.add(match)
        out.append(match)
    return out


def pick_offered_window(reply: str, offered: list[str]) -> str | None:
    if not (reply or "").strip() or not offered:
        return None
    lower = reply.lower()
    exact = [window for window in offered if window.lower() in lower]
    if len(exact) == 1:
        return exact[0]

    day_hits = []
    for window in offered:
        day = window.split()[0][:3].lower() if window.split() else ""
        aliases = _DAY_ALIASES.get(day, (day,))
        if any(re.search(rf"\b{re.escape(alias)}\b", reply, re.IGNORECASE) for alias in aliases):
            day_hits.append(window)
    if len(day_hits) == 1:
        return day_hits[0]

    time_hits = []
    for window in offered:
        parts = window.split()
        if len(parts) > 1 and parts[1].lower() in lower:
            time_hits.append(window)
    if len(time_hits) == 1:
        return time_hits[0]

    if re.search(r"\b(first|1st|earlier|the first one)\b", reply, re.IGNORECASE):
        return offered[0]
    if re.search(r"\b(second|2nd|later|the other|the second)\b", reply, re.IGNORECASE) and len(
        offered
    ) > 1:
        return offered[1]
    return None


def compose_first_touch_reply_sms(
    *,
    contact_first_name: str,
    agent_first_name: str,
    tone: str,
    windows: list[dict[str, Any]],
    picked: str | None = None,
    property_name: str | None = None,
    business_name: str | None = None,
) -> str:
    lead = first_name_of(contact_first_name, "there")
    who = first_name_of(agent_first_name, "") or (business_name or "").strip() or "I"
    tone_key = normalize_tone(tone)

    if picked:
        picked_label = picked.strip()
        if not picked_label:
            raise ValueError("first-touch reply draft is empty")
        if tone_key == "direct":
            text = f"Hi {lead} — {who} here. {picked_label} still work for you?"
        elif tone_key == "formal":
            text = f"Hello {lead}, this is {who}. Would {picked_label} still work for you?"
        elif tone_key == "casual":
            text = f"Hey {lead}! {who} here — {picked_label} still good?"
        else:
            text = f"Hey {lead}, this is {who}. I can do {picked_label} — does that still work?"
        content = " ".join(text.split()).strip()
        assert_valid_first_touch_reply_text(
            content, windows=[picked_label], agent_token=who, tone=tone_key, picked=picked_label
        )
        return content

    return compose_first_touch_sms(
        contact_first_name=contact_first_name,
        agent_first_name=agent_first_name,
        tone=tone,
        windows=windows,
        property_name=property_name,
        business_name=business_name,
    )


def assert_valid_first_touch_reply_text(
    content: str,
    *,
    windows: list[str],
    agent_token: str,
    tone: str,
    picked: str | None = None,
) -> None:
    if not content.strip():
        raise ValueError("first-touch reply draft is empty")
    lower = content.lower()
    if any(word in lower for word in ("sent", "delivered", "auto-sent", "autosent")):
        raise ValueError("first-touch reply draft claims it was sent")
    if any(
        word in lower
        for word in (
            "booked",
            "live",
            "reserved",
            "locked",
            "is held",
            "i'll lock",
            "i’ll lock",
            "see you then",
            "see you there",
        )
    ):
        raise ValueError("first-touch reply draft claims the showing is booked")
    if "chippy" in lower:
        raise ValueError("first-touch reply draft used the wrong brand spelling")
    if agent_token and agent_token not in content:
        raise ValueError("first-touch reply draft is not in the assigned agent voice")
    if picked:
        if picked not in content:
            raise ValueError(f"first-touch reply draft missing confirmed window: {picked}")
        marks = _CONFIRM_MARKS[normalize_tone(tone)]
        if not any(mark in lower for mark in marks):
            raise ValueError(f"first-touch reply draft is not in the {tone} voice")
        return
    if len(windows) < 2:
        raise ValueError("first-touch reply SMS requires two showing windows")
    for window in windows:
        if window not in content:
            raise ValueError(f"first-touch reply draft missing showing window: {window}")


def first_touch_reply_instruction(triggers: list[dict]) -> str | None:
    inbound = [
        t
        for t in triggers
        if is_inbound_message_event(t.get("event")) and t.get("contactId")
    ]
    if not inbound:
        return None
    lines = []
    for t in inbound:
        lines.append(f"- inbound_message contactId: {t['contactId']}")
    return (
        "FIRST-TOUCH REPLY — the lead answered the first-touch SMS. Send one "
        "short SMS in the assigned realtor's voice that books the showing: "
        "confirm the time they picked, or offer two concrete windows. Do not "
        "park a draft. Do not wait for approval.\n" + "\n".join(lines)
    )


def _property_from_contact(contact: dict[str, Any]) -> str | None:
    address = (contact.get("address") or "").strip()
    if address:
        return address
    for item in contact.get("properties") or []:
        if isinstance(item, str) and item.strip():
            return item.strip()
    data = contact.get("applicationData")
    if isinstance(data, dict):
        for key in (
            "address",
            "propertyAddress",
            "listingAddress",
            "property",
            "interestedProperty",
        ):
            value = data.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()
    return None


def _skipped(contact_id: str, reason: str) -> dict[str, Any]:
    return {
        "action": "skipped",
        "contactId": contact_id,
        "status": "skipped",
        "channel": "sms",
        "content": "",
        "windows": [],
        "sent": False,
        "reason": reason,
    }


async def ensure_first_touch_reply_draft(
    space_id: str,
    contact_id: str,
    *,
    reply_text: str | None = None,
    source_draft_id: str | None = None,
    channel: str | None = None,
    now: datetime | None = None,
) -> dict[str, Any]:
    """Send a first-touch-reply SMS. Persists sent. Never parks pending."""
    from db import supabase
    from tools.base import with_retry

    if channel and channel != "sms":
        return _skipped(contact_id, "not_sms")

    when = now or datetime.now(timezone.utc)
    db = await supabase()

    check = await (
        db.table("Contact")
        .select("id,name,phone,address,properties,applicationData")
        .eq("id", contact_id)
        .eq("spaceId", space_id)
        .maybe_single()
        .execute()
    )
    if not check.data:
        return {"error": "Contact not found in space", "sent": False}

    contact = check.data
    cutoff = (when - timedelta(days=_FIRST_TOUCH_LOOKBACK_DAYS)).isoformat()
    existing = await (
        db.table("AgentDraft")
        .select("id,content,status,channel,reasoning,createdAt")
        .eq("spaceId", space_id)
        .eq("contactId", contact_id)
        .eq("channel", "sms")
        .gte("createdAt", cutoff)
        .order("createdAt", desc=True)
        .limit(20)
        .execute()
    )
    drafts = existing.data or []
    source = next((row for row in drafts if row.get("id") == source_draft_id), None)
    first_touch = (
        source
        if source and source.get("status") != "dismissed"
        else next(
            (
                row
                for row in drafts
                if is_first_touch_draft(row) and row.get("status") != "dismissed"
            ),
            None,
        )
    )
    if not first_touch:
        return _skipped(contact_id, "no_first_touch")

    reply_cutoff = when - timedelta(hours=_DEDUPE_WINDOW_HOURS)

    def _in_window(row: dict[str, Any]) -> bool:
        created = row.get("createdAt")
        if not created:
            return True
        try:
            created_at = datetime.fromisoformat(created.replace("Z", "+00:00"))
            return created_at >= reply_cutoff
        except ValueError:
            return True

    already_sent = next(
        (
            row
            for row in drafts
            if is_first_touch_reply_draft(row)
            and row.get("status") == "sent"
            and (row.get("content") or "").strip()
            and _in_window(row)
        ),
        None,
    )
    empty_stub = next(
        (
            row
            for row in drafts
            if is_first_touch_reply_draft(row)
            and not (row.get("content") or "").strip()
            and _in_window(row)
        ),
        None,
    )

    if already_sent:
        return {
            "action": "deduped",
            "draftId": already_sent["id"],
            "status": "sent",
            "channel": "sms",
            "content": already_sent["content"],
            "sent": True,
        }

    profile_res = await (
        db.table("AIUserProfile")
        .select("displayName,communicationTone")
        .eq("spaceId", space_id)
        .maybe_single()
        .execute()
    )
    setting_res = await (
        db.table("SpaceSetting")
        .select("businessName,timezone,tourStartHour,tourEndHour,tourDaysAvailable")
        .eq("spaceId", space_id)
        .maybe_single()
        .execute()
    )
    space_res = await (
        db.table("Space").select("name").eq("id", space_id).maybe_single().execute()
    )
    profile = profile_res.data or {}
    setting = setting_res.data or {}
    space = space_res.data or {}

    agent_name = first_name_of(profile.get("displayName"), "")
    business = (setting.get("businessName") or space.get("name") or "").strip() or None
    if not agent_name:
        agent_name = business or "I"

    offered = extract_window_labels(first_touch.get("content") or "")
    picked = pick_offered_window(reply_text or "", offered)
    if picked:
        windows = [{"label": picked}]
    else:
        windows = propose_two_showing_windows(
            when,
            start_hour=setting.get("tourStartHour") or 9,
            end_hour=setting.get("tourEndHour") or 17,
            days_available=setting.get("tourDaysAvailable") or None,
        )
        if len(windows) < 2:
            return {"error": "could not propose two showing windows", "sent": False}

    content = compose_first_touch_reply_sms(
        contact_first_name=first_name_of(contact.get("name"), "there"),
        agent_first_name=agent_name,
        tone=normalize_tone(profile.get("communicationTone")),
        windows=windows,
        picked=picked,
        property_name=_property_from_contact(contact),
        business_name=business,
    )
    if not content.strip():
        raise ValueError("first-touch reply draft is empty")

    await send_sms(to=contact.get("phone"), body=content, label="first-touch-reply")

    expires_at = (when + timedelta(days=7)).isoformat()
    if empty_stub:
        await with_retry(
            lambda: db.table("AgentDraft")
            .update(
                {
                    "content": content,
                    "reasoning": _REPLY_REASON,
                    "priority": 85,
                    "status": "sent",
                    "expiresAt": expires_at,
                    "updatedAt": when.isoformat(),
                }
            )
            .eq("id", empty_stub["id"])
            .eq("spaceId", space_id)
            .execute()
        )
        return {
            "action": "filled",
            "draftId": empty_stub["id"],
            "status": "sent",
            "channel": "sms",
            "content": content,
            "windows": [w["label"] for w in windows],
            "picked": picked,
            "sent": True,
        }

    draft = {
        "id": str(uuid.uuid4()),
        "spaceId": space_id,
        "contactId": contact_id,
        "channel": "sms",
        "content": content,
        "reasoning": _REPLY_REASON,
        "priority": 85,
        "status": "sent",
        "expiresAt": expires_at,
    }
    result = await with_retry(lambda: db.table("AgentDraft").insert(draft).execute())
    created = result.data[0] if result.data else draft
    return {
        "action": "sent",
        "draftId": created.get("id", draft["id"]),
        "status": "sent",
        "channel": "sms",
        "content": content,
        "windows": [w["label"] for w in windows],
        "picked": picked,
        "sent": True,
    }

File: crm/lib/agent/first-touch-reply.ts
/**
 * Booking SMS after a lead replies to first-touch.
 *
 * First-touch offered two windows. When the lead texts back, Chippi
 * writes the next message that books the showing — confirm the time they
 * picked, or put two concrete windows in front of them again — and sends
 * it through Telnyx. No approval inbox. No pending persist. Missing
 * credentials fail.
 *
 * Called from the inbound_message trigger so the SMS goes out even if
 * Modal is slow, and again from the autonomous run as a backstop.
 */

import { supabase } from '@/lib/supabase';
import { logger } from '@/lib/logger';
import {
  composeFirstTouchSms,
  firstNameOf,
  normalizeTone,
  proposeTwoShowingWindows,
  sendAutonomousSms,
  type AgentTone,
  type FirstTouchVoice,
  type ShowingWindow,
} from '@/lib/agent/first-touch';
import { isInboundMessageEvent } from '@/lib/agent/trigger-policy';

export type FirstTouchReplyAction = 'sent' | 'deduped' | 'filled' | 'skipped';

export interface FirstTouchReplyDraftResult {
  action: FirstTouchReplyAction;
  draftId?: string;
  contactId: string;
  channel: 'sms';
  status: 'sent' | 'skipped';
  content: string;
  windows: string[];
  picked?: string;
  sent: boolean;
  reason?: string;
}

export interface DraftFirstTouchReplyInput {
  spaceId: string;
  contactId: string;
  replyText?: string;
  sourceDraftId?: string;
  channel?: 'sms' | 'email';
  now?: Date;
}

const DEDUPE_WINDOW_HOURS = 48;
const FIRST_TOUCH_LOOKBACK_DAYS = 14;
const DEFAULT_TIMEZONE = 'America/New_York';
const DEFAULT_START_HOUR = 9;
const DEFAULT_END_HOUR = 17;
const DEFAULT_DAYS = [1, 2, 3, 4, 5];

const FIRST_TOUCH_REASON_MARK = 'First-touch SMS';
const REPLY_REASON = 'Reply to first-touch — book a showing. Sent.';

const WINDOW_LABEL_RE =
  /\b(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)\s+\d{1,2}(?::\d{2})?(?:am|pm)\b/gi;

const DAY_ALIASES: Record<string, string[]> = {
  mon: ['mon', 'monday'],
  tue: ['tue', 'tues', 'tuesday'],
  wed: ['wed', 'wednesday'],
  thu: ['thu', 'thur', 'thurs', 'thursday'],
  fri: ['fri', 'friday'],
  sat: ['sat', 'saturday'],
  sun: ['sun', 'sunday'],
};

export function isInboundFirstTouchReplyEvent(event: string): boolean {
  return isInboundMessageEvent(event);
}

export function isFirstTouchDraft(row: { reasoning?: string | null }): boolean {
  return (row.reasoning ?? '').includes(FIRST_TOUCH_REASON_MARK);
}

export function isFirstTouchReplyDraft(row: { reasoning?: string | null }): boolean {
  return (row.reasoning ?? '').includes('Reply to first-touch');
}

export function extractWindowLabels(text: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const match of text.match(WINDOW_LABEL_RE) ?? []) {
    if (seen.has(match)) continue;
    seen.add(match);
    out.push(match);
  }
  return out;
}

export function pickOfferedWindow(reply: string, offered: string[]): string | undefined {
  if (!reply.trim() || offered.length === 0) return undefined;
  const lower = reply.toLowerCase();

  const exact = offered.filter((window) => lower.includes(window.toLowerCase()));
  if (exact.length === 1) return exact[0];

  const dayHits = offered.filter((window) => {
    const day = window.split(/\s+/)[0]?.toLowerCase().slice(0, 3);
    const aliases = DAY_ALIASES[day] ?? [day];
    return aliases.some((alias) => new RegExp(`\\b${alias}\\b`, 'i').test(reply));
  });
  if (dayHits.length === 1) return dayHits[0];

  const timeHits = offered.filter((window) => {
    const time = window.split(/\s+/)[1];
    return time && lower.includes(time.toLowerCase());
  });
  if (timeHits.length === 1) return timeHits[0];

  if (/\b(first|1st|earlier|the first one)\b/i.test(reply) && offered[0]) {
    return offered[0];
  }
  if (/\b(second|2nd|later|the other|the second)\b/i.test(reply) && offered[1]) {
    return offered[1];
  }
  return undefined;
}

export function composeFirstTouchReplySms(input: {
  contactFirstName: string;
  voice: FirstTouchVoice;
  windows: ShowingWindow[];
  picked?: string;
  property?: string;
}): string {
  const lead = firstNameOf(input.contactFirstName, 'there');
  const agent = firstNameOf(input.voice.agentFirstName, '');
  const who = agent || input.voice.businessName?.trim() || 'I';

  if (input.picked) {
    const picked = input.picked.trim();
    if (!picked) {
      throw new Error('first-touch reply draft is empty');
    }
    let text: string;
    switch (input.voice.tone) {
      case 'direct':
        text = `Hi ${lead} — ${who} here. ${picked} still work for you?`;
        break;
      case 'formal':
        text = `Hello ${lead}, this is ${who}. Would ${picked} still work for you?`;
        break;
      case 'casual':
        text = `Hey ${lead}! ${who} here — ${picked} still good?`;
        break;
      case 'warm':
      default:
        text = `Hey ${lead}, this is ${who}. I can do ${picked} — does that still work?`;
        break;
    }
    const content = text.replace(/\s+/g, ' ').trim();
    assertValidFirstTouchReplyText(content, {
      windows: [picked],
      agentToken: who,
      tone: input.voice.tone,
      picked,
    });
    return content;
  }

  return composeFirstTouchSms({
    contactFirstName: input.contactFirstName,
    voice: input.voice,
    windows: input.windows,
    property: input.property,
  });
}

export function assertValidFirstTouchReplyText(
  content: string,
  opts: { windows: string[]; agentToken: string; tone: AgentTone; picked?: string },
): void {
  if (!content.trim()) {
    throw new Error('first-touch reply draft is empty');
  }
  if (/\b(sent|delivered|auto-?sent)\b/i.test(content)) {
    throw new Error('first-touch reply draft claims it was sent');
  }
  if (
    /\b(booked|live|reserved|locked)\b/i.test(content) ||
    /\bis held\b/i.test(content) ||
    /\bi['’]ll lock\b/i.test(content) ||
    /\bsee you (then|there)\b/i.test(content)
  ) {
    throw new Error('first-touch reply draft claims the showing is booked');
  }
  if (/\bchippy\b/i.test(content)) {
    throw new Error('first-touch reply draft used the wrong brand spelling');
  }
  if (opts.agentToken && !content.includes(opts.agentToken)) {
    throw new Error('first-touch reply draft is not in the assigned agent voice');
  }
  if (opts.picked) {
    if (!content.includes(opts.picked)) {
      throw new Error(`first-touch reply draft missing confirmed window: ${opts.picked}`);
    }
    const confirmMarks: Record<AgentTone, RegExp> = {
      warm: /does that still work/i,
      direct: /still work for you/i,
      formal: /would .+ still work/i,
      casual: /still good/i,
    };
    if (!confirmMarks[opts.tone].test(content)) {
      throw new Error(`first-touch reply draft is not in the ${opts.tone} voice`);
    }
    return;
  }
  if (opts.windows.length < 2) {
    throw new Error('first-touch reply SMS requires two showing windows');
  }
  for (const window of opts.windows) {
    if (!content.includes(window)) {
      throw new Error(`first-touch reply draft missing showing window: ${window}`);
    }
  }
}

function propertyFromContact(contact: {
  address?: string | null;
  properties?: string[] | null;
  applicationData?: unknown;
}): string | undefined {
  if (contact.address?.trim()) return contact.address.trim();
  const listed = (contact.properties ?? []).find((p) => typeof p === 'string' && p.trim());
  if (listed) return listed.trim();
  const data = contact.applicationData;
  if (data && typeof data === 'object') {
    const rec = data as Record<string, unknown>;
    for (const key of ['address', 'propertyAddress', 'listingAddress', 'property', 'interestedProperty']) {
      const value = rec[key];
      if (typeof value === 'string' && value.trim()) return value.trim();
    }
  }
  return undefined;
}

type DraftRow = {
  id: string;
  content: string | null;
  status: string;
  channel: string;
  reasoning?: string | null;
  createdAt?: string;
};

function skipped(
  contactId: string,
  reason: string,
): FirstTouchReplyDraftResult {
  return {
    action: 'skipped',
    contactId,
    channel: 'sms',
    status: 'skipped',
    content: '',
    windows: [],
    sent: false,
    reason,
  };
}

export async function draftFirstTouchReplyForLead(
  input: DraftFirstTouchReplyInput,
): Promise<FirstTouchReplyDraftResult> {
  if (!input.spaceId || !input.contactId) {
    throw new Error('spaceId and contactId are required');
  }
  if (input.channel && input.channel !== 'sms') {
    return skipped(input.contactId, 'not_sms');
  }

  const now = input.now ?? new Date();
  const { data: contact, error: contactError } = await supabase
    .from('Contact')
    .select('id,name,phone,address,properties,applicationData,spaceId')
    .eq('id', input.contactId)
    .eq('spaceId', input.spaceId)
    .maybeSingle();

  if (contactError) {
    logger.error('[first-touch-reply] contact lookup failed', { spaceId: input.spaceId }, contactError);
    throw new Error('Contact lookup failed');
  }
  if (!contact) {
    throw new Error('Contact not found in space');
  }

  const cutoff = new Date(now.getTime() - FIRST_TOUCH_LOOKBACK_DAYS * 24 * 60 * 60 * 1000).toISOString();
  const { data: existingRows, error: draftsError } = await supabase
    .from('AgentDraft')
    .select('id,content,status,channel,reasoning,createdAt')
    .eq('spaceId', input.spaceId)
    .eq('contactId', input.contactId)
    .eq('channel', 'sms')
    .gte('createdAt', cutoff)
    .order('createdAt', { ascending: false })
    .limit(20);
  if (draftsError) {
    logger.error('[first-touch-reply] draft lookup failed', { spaceId: input.spaceId }, draftsError);
    throw new Error('Draft lookup failed');
  }

  const drafts = (existingRows ?? []) as DraftRow[];
  const source = input.sourceDraftId
    ? drafts.find((row) => row.id === input.sourceDraftId && isFirstTouchDraft(row))
    : undefined;
  const firstTouch =
    (source && source.status !== 'dismissed' ? source : undefined) ??
    drafts.find((row) => isFirstTouchDraft(row) && row.status !== 'dismissed');

  if (!firstTouch) {
    return skipped(input.contactId, 'no_first_touch');
  }

  const replyCutoffMs = now.getTime() - DEDUPE_WINDOW_HOURS * 60 * 60 * 1000;
  const alreadySent = drafts.find((row) => {
    if (!isFirstTouchReplyDraft(row) || row.status !== 'sent' || !row.content?.trim()) return false;
    if (!row.createdAt) return true;
    return new Date(row.createdAt).getTime() >= replyCutoffMs;
  });
  const emptyStub = drafts.find((row) => {
    if (!isFirstTouchReplyDraft(row) || Boolean(row.content?.trim())) return false;
    if (!row.createdAt) return true;
    return new Date(row.createdAt).getTime() >= replyCutoffMs;
  });

  const [profileRes, settingRes, spaceRes] = await Promise.all([
    supabase
      .from('AIUserProfile')
      .select('displayName,communicationTone')
      .eq('spaceId', input.spaceId)
      .maybeSingle(),
    supabase
      .from('SpaceSetting')
      .select('businessName,timezone,tourStartHour,tourEndHour,tourDaysAvailable')
      .eq('spaceId', input.spaceId)
      .maybeSingle(),
    supabase.from('Space').select('name').eq('id', input.spaceId).maybeSingle(),
  ]);

  const profile = profileRes.data as { displayName?: string | null; communicationTone?: string | null } | null;
  const setting = settingRes.data as {
    businessName?: string | null;
    timezone?: string | null;
    tourStartHour?: number | null;
    tourEndHour?: number | null;
    tourDaysAvailable?: number[] | null;
  } | null;
  const space = spaceRes.data as { name?: string | null } | null;

  const voice: FirstTouchVoice = {
    tone: normalizeTone(profile?.communicationTone),
    agentFirstName: firstNameOf(profile?.displayName, ''),
    businessName: setting?.businessName?.trim() || space?.name?.trim() || undefined,
  };
  if (!voice.agentFirstName && voice.businessName) {
    voice.agentFirstName = voice.businessName;
  }
  if (!voice.agentFirstName) {
    voice.agentFirstName = 'I';
  }

  const offered = extractWindowLabels(firstTouch.content ?? '');
  const picked = pickOfferedWindow(input.replyText ?? '', offered);
  const windows = picked
    ? [{ startsAt: now, label: picked }]
    : proposeTwoShowingWindows(now, {
        timeZone: setting?.timezone ?? DEFAULT_TIMEZONE,
        startHour: setting?.tourStartHour ?? DEFAULT_START_HOUR,
        endHour: setting?.tourEndHour ?? DEFAULT_END_HOUR,
        daysAvailable: setting?.tourDaysAvailable ?? DEFAULT_DAYS,
      });
  if (!picked && windows.length < 2) {
    throw new Error('could not propose two showing windows');
  }

  const content = composeFirstTouchReplySms({
    contactFirstName: firstNameOf(contact.name, 'there'),
    voice,
    windows,
    picked,
    property: propertyFromContact(contact),
  });
  if (!content.trim()) {
    throw new Error('first-touch reply draft is empty');
  }

  const windowLabels = picked ? [picked] : windows.map((w) => w.label);

  if (alreadySent) {
    return {
      action: 'deduped',
      draftId: alreadySent.id,
      contactId: input.contactId,
      channel: 'sms',
      status: 'sent',
      content: alreadySent.content!,
      windows: windowLabels,
      picked,
      sent: true,
    };
  }

  await sendAutonomousSms({
    to: (contact as { phone?: string | null }).phone,
    body: content,
    label: 'first-touch-reply',
  });

  const expiresAt = new Date(now.getTime() + 7 * 24 * 60 * 60 * 1000).toISOString();
  const row = {
    id: emptyStub?.id ?? crypto.randomUUID(),
    spaceId: input.spaceId,
    contactId: input.contactId,
    channel: 'sms' as const,
    content,
    reasoning: REPLY_REASON,
    priority: 85,
    status: 'sent' as const,
    expiresAt,
    updatedAt: now.toISOString(),
  };

  if (emptyStub) {
    const { data: updated, error: updateError } = await supabase
      .from('AgentDraft')
      .update({
        content: row.content,
        reasoning: row.reasoning,
        priority: row.priority,
        status: 'sent',
        expiresAt: row.expiresAt,
        updatedAt: row.updatedAt,
      })
      .eq('id', emptyStub.id)
      .eq('spaceId', input.spaceId)
      .select('id');
    if (!updateError && updated && updated.length > 0) {
      return {
        action: 'filled',
        draftId: emptyStub.id,
        contactId: input.contactId,
        channel: 'sms',
        status: 'sent',
        content,
        windows: windowLabels,
        picked,
        sent: true,
      };
    }
    if (updateError) {
      logger.error('[first-touch-reply] failed to record sent SMS', { spaceId: input.spaceId }, updateError);
    }
    row.id = crypto.randomUUID();
  }

  const { data: inserted, error: insertError } = await supabase
    .from('AgentDraft')
    .insert(row)
    .select('id')
    .maybeSingle();
  if (insertError) {
    logger.error('[first-touch-reply] insert failed', { spaceId: input.spaceId }, insertError);
    throw new Error('Failed to record first-touch reply SMS');
  }

  return {
    action: 'sent',
    draftId: (inserted as { id?: string } | null)?.id ?? row.id,
    contactId: input.contactId,
    channel: 'sms',
    status: 'sent',
    content,
    windows: windowLabels,
    picked,
    sent: true,
  };
}
